sokol_audio.h

```diff
@@ -2282,7 +2346,151 @@ _SOKOL_PRIVATE void _saudio_vita_backend_shutdown(void) {
+        if (defaultUserData == NULL) {
+            _SAUDIO_ERROR(N3DS_USERDATA_DEFAULT_MALLOC_FAILED);
+            return false;
+        }
+
+        defaultUserData->queue_count = 2;
+        defaultUserData->interpolation_type = NDSP_INTERP_POLYPHASE;
+        defaultUserData->channel_id = 0;
+
+        _saudio.desc.user_data = defaultUserData;
+    }
+
+    user_data_3ds_t* user_data = (user_data_3ds_t*)_saudio.desc.user_data;
+
+    /* clamp to 2 channels max */
+    if (_saudio.num_channels > 2) {
+        _saudio.num_channels = 2;
+    }
+
+    ndspChnReset(user_data->channel_id);
+    ndspChnWaveBufClear(user_data->channel_id);
+    ndspChnSetInterp(user_data->channel_id, user_data->interpolation_type);
+    ndspChnSetRate(user_data->channel_id, _saudio.sample_rate);
+    ndspChnSetFormat(user_data->channel_id, _saudio.num_channels == 1 ? NDSP_FORMAT_MONO_PCM16 : NDSP_FORMAT_STEREO_PCM16);
+    ndspSetOutputMode(_saudio.num_channels == 1 ? NDSP_OUTPUT_MONO : NDSP_OUTPUT_STEREO);
+
+    /* read back actual sample rate and channels */
+    _saudio.sample_rate = (int)_saudio.sample_rate;
+    _saudio.bytes_per_frame = _saudio.num_channels * (int)sizeof(float);
+
+    /* allocate the streaming buffer */
+    _saudio.backend.buffer_byte_size = _saudio.buffer_frames * _saudio.bytes_per_frame;
+    _saudio.backend.buffer = (float*) _saudio_malloc_clear((size_t)_saudio.backend.buffer_byte_size);
+    _saudio.backend.buffer_3ds = (int16_t*)linearAlloc(user_data->queue_count * _saudio.buffer_frames * _saudio.num_channels * sizeof(int16_t));
+    _saudio.backend.queue_3ds = (ndspWaveBuf*)_saudio_malloc(user_data->queue_count * sizeof(ndspWaveBuf));
+
+    /* prepare the 3ds audio queues */
+    int16_t* bufferPtrCopy = _saudio.backend.buffer_3ds;
+    for (int i = 0; i < user_data->queue_count; ++i) {
+        _saudio.backend.queue_3ds[i].data_vaddr = bufferPtrCopy; /* point the queue at the section of the linear buffer */
+        _saudio.backend.queue_3ds[i].looping = false; /* the user should handle looping on their end */
+        _saudio.backend.queue_3ds[i].status = NDSP_WBUF_DONE; /* default to done status for buffering logic */
+
+        bufferPtrCopy += _saudio.buffer_frames * _saudio.num_channels;
+    }
+
+    /* misc settings */
+    _saudio.backend.channel_id = user_data->channel_id;
+
+    /* instead of a thread, ndsp will trigger a callback */
+    /* when it needs more data.                          */
+    ndspSetCallback(_saudio_3ds_ndsptrigger_cb, NULL);
+
+    return true;
+}
+
+_SOKOL_PRIVATE void _saudio_3ds_backend_shutdown(void) {
+    _saudio.backend.thread_stop = true;
+
+    if (_saudio.backend.channel_id >= 0) {
+        ndspChnWaveBufClear(_saudio.backend.channel_id);
+        ndspExit();
+        _saudio.backend.channel_id = -1;
+    }
+
+    _saudio_free(_saudio.backend.queue_3ds);
+    _saudio_free(_saudio.backend.buffer_3ds);
+    _saudio_free(_saudio.backend.buffer);
```

Comment:
I think you'd need to also free the optionally allocated user-data block here, but as I said above, those config params should just go into nested structs instead :)